### PR TITLE
import pydantic objects from the `_pydantic_compat` module

### LINKED ISSUE #17659: `check_pydantic_models.py` should ignore external modules
The [check_pydantic_models.py](https://github.com/element-hq/synapse/blob/develop/scripts-dev/check_pydantic_models.py) script patches all pydantic imports, including the imports from external modules. #17144 introduces a dependency to [scim2-models](https://scim2-models.readthedocs.io) that requires [pydantic >= 2.7](https://github.com/yaal-coop/scim2-models/blob/9fb4ed1bc5ace168e661023b512880bc99d5155e/pyproject.toml#L35) and [breaks the script](https://github.com/element-hq/synapse/pull/17144#issuecomment-2297336639).

After discussing it on #synapse-dev, it seems `check_pydantic_models.py` should ignore external modules.

Related to https://github.com/matrix-org/synapse/pull/13502, https://github.com/matrix-org/synapse/pull/16332

### PR DESCRIPTION
**Pull request by azmeuk**
This PR changes `from pydantic import BaseModel` to `from synapse._pydantic_compat import BaseModel` (as well as `constr`, `conbytes`, `conint`, `confloat`).

It allows `check_pydantic_models.py` to mock those pydantic objects only in the synapse module, and not interfere with pydantic objects in external dependencies.

This should solve the CI problems for #17144, which breaks because `check_pydantic_models.py` patches pydantic models from [scim2-models](https://scim2-models.readthedocs.io/).

/cc @DMRobertson @gotmax23
fixes #17659 

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/types/rest/client/__init__.py [hunk 1]
```diff
@@ -384,7 +380,7 @@ class TypingExtension(RequestBodyModel):
         receipts: Optional[ReceiptsExtension] = None
         typing: Optional[TypingExtension] = None
 
-    conn_id: Optional[str]
```
Comment (azmeuk) [inline synapse/types/rest/client/__init__.py:387]: My modifications revealed an issue with this `str` type that was rejected by `check_pydantic_models`. I am not sure if this was wanted though.

File: synapse/config/_util.py [hunk 2]
```diff
@@ -22,12 +22,12 @@
 
 import jsonschema
 
-from synapse._pydantic_compat import HAS_PYDANTIC_V2
+from synapse._pydantic_compat import HAS_PYDANTIC_V2, BaseModel
 
 if TYPE_CHECKING or HAS_PYDANTIC_V2:
-    from pydantic.v1 import BaseModel, ValidationError, parse_obj_as
+    from pydantic.v1 import ValidationError, parse_obj_as
```
Comment (gotmax23) [inline synapse/config/_util.py:28]: The other classes/functions should also be added to the `_pydantic_compat` module if you're going to do it like this. Importing some members directly from `pydantic` / `pydantic.v1` and others from the compat module is confusing.

Comment (azmeuk) [reply synapse/config/_util.py]: :+1: This is fixed now with 7a40450

File: synapse/types/rest/client/__init__.py (hunk 1 above)
Comment (anoadragon453) [reply synapse/types/rest/client/__init__.py:387]: This change seems fine to me. `conn_id` should be a string.

**Review by anoadragon453 — APPROVED** [review]
The new ergonomics of the imports are great. Thanks for this!

**Pull request merged**